bin/exec.ml

```diff
@@ -32,6 +32,75 @@ let man =
 
 let info = Cmd.info "exec" ~doc ~man
 
+module Command_to_exec = struct
+  (* A command to execute, which knows how to (re)build the program and then
+     run it with some arguments in an enivorment *)
+
+  type t =
+    { get_path_and_build_if_necessary :
+        unit -> (Path.t, [ `Already_reported ]) result Fiber.t
+    ; args : string list
+    ; env : Env.t
+    }
+
+  (* Helper function to spawn a new process running a command in an
+     environment, returning the new process' pid *)
+  let spawn_process path ~args ~env =
+    let path = Path.to_string path in
+    let env = Env.to_unix env |> Spawn.Env.of_list in
+    let argv = path :: args in
+    let pid = Spawn.spawn ~prog:path ~env ~argv () in
+    Pid.of_int pid
+
+  (* Run the command, first (re)building the program which the command is
+     invoking *)
+  let build_and_run_in_child_process
+      { get_path_and_build_if_necessary; args; env } =
+    get_path_and_build_if_necessary ()
+    |> Fiber.map ~f:(Result.map ~f:(spawn_process ~args ~env))
```

Comment:
We should release the global lock before running the process and then acquire it again. So that the user can start builds while their program is running.

Reply:
Updated to take the lock around running the build system